src/nevermined/Nfts.ts

```diff
@@ -512,4 +515,148 @@ export class Nfts extends Instantiable {
+
+        if (saveResult) {
+            return agreementId
+        } else {
+            throw Error(`Error saving ${agreementId} to MetadataDB`)
+        }
+    }
+
+    /**
+     * Buys a number of listed goods on secondary markets.
+     * @param buyer The address of the buyer.
+     * @param seller The address of the seller.
+     * @param numOfNfts The number of assets to buy. 1 by default.
+     * @param ddo The DDO of the NFT.
+     * @param provider The provider address.
+     * @param agreementId The agreementId of the initial sales agreement created off-chain.
+     * @returns true if the buy was successful.
+     */
+    public async buySecondaryMarketNft(
+        buyer: string,
+        seller: string,
+        numOfNfts: number = 1,
+        ddo: DDO,
+        provider: string,
+        agreementId: string
+    ): Promise<boolean> {
+        const buyerAccount = new Account(buyer)
+        const sellerAccount = new Account(seller)
+        const { nftSalesTemplate } = this.nevermined.keeper.templates
+        let providerAccounts: Account
+        if (provider && provider !== null && provider !== '') {
+            providerAccounts = new Account(provider)
+        }
+        const service = await this.nevermined.metadata.retrieveServiceAgreement(
+            agreementId
+        )
+        const assetRewards = getAssetRewardsFromService(service)
+
+        const result = await nftSalesTemplate.createAgreementFromDDO(
+            agreementId,
+            ddo,
+            assetRewards,
+            buyerAccount,
+            numOfNfts,
+            providerAccounts || sellerAccount,
+            buyerAccount,
+            service as TxParameters
+        )
+
+        if (!result) throw new Error('Creating buy agreement failed')
+
+        const payment = findServiceConditionByName(service, 'lockPayment')
+
+        let receipt = await this.nevermined.agreements.conditions.lockPayment(
+            agreementId,
+            ddo.id,
+            assetRewards.getAmounts(),
+            assetRewards.getReceivers(),
+            payment.parameters.find(p => p.name === '_tokenAddress').value as string,
+            buyerAccount
+        )
+
+        if (!receipt) throw new Error('LockPayment Failed.')
+
+        receipt = await this.nevermined.agreements.conditions.transferNft(
+            agreementId,
+            ddo,
+            assetRewards.getAmounts(),
+            assetRewards.getReceivers(),
+            numOfNfts,
+            sellerAccount,
+            assetRewards as TxParameters
+        )
+
+        if (!receipt) throw new Error('TranferNft Failed.')
+
+        receipt = await this.nevermined.agreements.conditions.releaseNftReward(
+            agreementId,
+            ddo,
+            assetRewards.getAmounts(),
+            assetRewards.getReceivers(),
+            numOfNfts,
+            sellerAccount
+        )
+
+        if (!receipt) throw new Error('ReleaseBftReward Failed.')
+        return true
+    }
```

Comment:
The transfer of the NFT and the release of the rewards cannot be done by the buyer it needs to be done by the seller. Somehow the seller needs to be notified of the lock payment and trigger the transfer and reward release

Reply:
That's what I feared. Either way, we need a third step & then change the whole flow to an auction based one. Or, can we delegate to the gateway?

Reply:
We can update the transferForDelegate on the gateway to support this but this is a bit dangerous because anyone can create a service agreement

Reply:
Yep, that's exactly what we were discussing. Most likely this functionality will need to be rethought on the marketplace levels. To do it via the gateway the only secure way is to allow creation of agreements only for those assets that have already been `put up for secondary sales`.